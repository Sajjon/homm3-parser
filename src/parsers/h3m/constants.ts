export const H3M_MAX_PLAYERS = 8

export enum MapFormat {
  ROE = 'RoE',
  AB = 'AB',
  SOD = 'SoD',
  CHR = 'Chr',
  WOG = 'WoG',
}

export const formatEnum = {
  [MapFormat.ROE]: 0x0000000e,
  [MapFormat.AB]: 0x00000015,
  [MapFormat.SOD]: 0x0000001c,
  [MapFormat.CHR]: 0x0000001d,
  [MapFormat.WOG]: 0x00000033,
}

export enum Difficulty {
  Easy = 'Easy',
  Normal = 'Normal',
  Hard = 'Hard',
  Expert = 'Expert',
  Impossible = 'Impossible',
}

export const difficultyEnum = {
  [Difficulty.Easy]: 0,
  [Difficulty.Normal]: 1,
  [Difficulty.Hard]: 2,
  [Difficulty.Expert]: 3,
  [Difficulty.Impossible]: 4,
}

export enum Behavior {
  Random = 'Random',
  Warrior = 'Warrior',
  Builder = 'Builder',
  Explorer = 'Explorer',
}

export const behaviorEnum = {
  [Behavior.Random]: 0,
  [Behavior.Warrior]: 1,
  [Behavior.Builder]: 2,
  [Behavior.Explorer]: 3,
}

export enum HallLevel {
  Town = 'Town',
  City = 'City',
  Capitol = 'Capitol',
}

export const hallLevelEnum = {
  [HallLevel.Town]: 0,
  [HallLevel.City]: 1,
  [HallLevel.Capitol]: 2,
}

export enum CastleLevel {
  Fort = 'Fort',
  Citadel = 'Citadel',
  Castle = 'Castle',
}

export const castleLevelEnum = {
  [CastleLevel.Fort]: 0,
  [CastleLevel.Citadel]: 1,
  [CastleLevel.Castle]: 2,
}

export enum Gender {
  Default = 'Default',
  Male = 'Male',
  Female = 'Female',
}

export const genderEnum = {
  [Gender.Default]: 0xff,
  [Gender.Male]: 0x00,
  [Gender.Female]: 0x01,
}
