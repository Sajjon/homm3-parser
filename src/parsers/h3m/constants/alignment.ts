export enum Alignment {
  Random = 'Random',
  Castle = 'Castle',
  Rampart = 'Rampart',
  Tower = 'Tower',
  Inferno = 'Inferno',
  Necropolis = 'Necropolis',
  Dungeon = 'Dungeon',
  Stronghold = 'Stronghold',
  Fortress = 'Fortress',
  Conflux = 'Conflux',
}

export const alignmentEnum = {
  [Alignment.Random]: -1,
  [Alignment.Castle]: 0,
  [Alignment.Rampart]: 1,
  [Alignment.Tower]: 2,
  [Alignment.Inferno]: 3,
  [Alignment.Necropolis]: 4,
  [Alignment.Dungeon]: 5,
  [Alignment.Stronghold]: 6,
  [Alignment.Fortress]: 7,
  [Alignment.Conflux]: 8,
}
