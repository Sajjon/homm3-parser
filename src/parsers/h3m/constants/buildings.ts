export enum Building {
  VillageHall,
  TownHall,
  CityHall,
  Capitol,
  Fort,
  Citadel,
  Castle,
  Tavern,
  Blacksmith,
  Marketplace,
  ResourceSilo,
  ArtifactMerchants,
  MageGuild1,
  MageGuild2,
  MageGuild3,
  MageGuild4,
  MageGuild5,
  Shipyard,
  Grail,
  Special1,
  Special2,
  Special3,
  Special4,
  Dwelling1,
  Horde1,
  DwellingUp1,
  Dwelling2,
  Horde2,
  DwellingUp2,
  Dwelling3,
  Horde3,
  DwellingUp3,
  Dwelling4,
  Horde4,
  DwellingUp4,
  Dwelling5,
  Horde5,
  DwellingUp5,
  Dwelling6,
  DwellingUp6,
  Dwelling7,
  DwellingUp7
}
