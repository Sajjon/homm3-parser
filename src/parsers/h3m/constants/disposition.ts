export enum Disposition {
  Compliant, // will always join hero
  Friendly, // likely to join hero
  Aggressive, // may join hero
  Hostile, // unlikely to join hero
  Savage // will never join hero
}

export const dispositioEnum = {
  [Disposition.Compliant]: 0,
  [Disposition.Friendly]: 1,
  [Disposition.Aggressive]: 2,
  [Disposition.Hostile]: 3,
  [Disposition.Savage]: 4
}
