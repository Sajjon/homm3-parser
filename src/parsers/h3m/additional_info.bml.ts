import {
  byte,
  uint8,
  uint32,
  branch,
  array,
  when,
  struct,
  ctx,
  bytes,
  enums,
  flag,
  skip
} from 'binary-markup'
import { H3M_MAX_PLAYERS, Gender } from './constants'
import {
  isRoE,
  isNotRoE,
  isSoD,
  primarySkills,
  secondarySkill,
  artifacts,
  hero,
  hommString,
  PrimarySkills,
  Artifacts,
  SecondarySkill
} from './common.bml'
import { LoseCondition, loseConditionsEnum } from './constants/lose'
import { LoseConditionData, loseConditionsBranches } from './additional_info/lose_conditions.bml'
import { WinCondition, winConditionsEnum } from './constants/win'
import { winConditionsBranches, WinConditionData } from './additional_info/win_conditions.bml'
import { Hero } from './constants/hero'
import { FlaggedProp } from '../../helpers/types'

export interface Rumor {
  name: string
  desciption: string
}
const rumor = struct(
  //
  hommString`name`,
  hommString`desciption`
)

export interface CustomHero {
  hero: Hero
  face: number
  name: string
  allowed_players: number
}
const customHero = struct(
  //
  hero`hero`,
  uint8`face`,
  hommString`name`,
  uint8`allowed_players`
)

export interface SecondarySkills {
  count: number
  entries: SecondarySkill[]
}
const secondarySkills = struct(
  //
  uint32`count`,
  array(secondarySkill, ctx`count`)`entries`
)

type WithExperience = FlaggedProp<'hasExperience', 'experience', number>
type WithSecondarySkills = FlaggedProp<'hasSecondarySkills', 'secondarySkills', SecondarySkills>
type WithArtifacts = FlaggedProp<'hasArtifacts', 'artifacts', Artifacts>
type WithBiography = FlaggedProp<'hasBiography', 'biography', string>
type WithSpells = FlaggedProp<'hasSpells', 'spells', number[]>
type WithPrimarySkills = FlaggedProp<'hasPrimarySkills', 'primarySkills', PrimarySkills>

export type HeroData = {
  gender: Gender
} & WithExperience &
  WithSecondarySkills &
  WithArtifacts &
  WithBiography &
  WithSpells &
  WithPrimarySkills

const heroSetting = struct(
  flag`hasExperience`,
  when(ctx`hasExperience`, uint32)`experience`,
  flag`hasSecondarySkills`,
  when(ctx`hasSecondarySkills`, secondarySkills)`secondarySkills`,
  flag`hasArtifacts`,
  when(ctx`hasArtifacts`, artifacts)`artifacts`,
  flag`hasBiography`,
  when(ctx`hasBiography`, hommString)`biography`,
  enums(uint8, { [Gender.Default]: 0xff, [Gender.Male]: 0x00, [Gender.Female]: 0x01 })`gender`,
  flag`hasSpells`,
  when(ctx`hasSpells`, bytes(9))`spells`,
  flag`hasPrimarySkills`,
  when(ctx`hasPrimarySkills`, primarySkills)`primarySkills`
)

export type HeroSettings = FlaggedProp<'hasSettings', 'settings', HeroData>

const heroSettings = struct(
  //
  flag`hasSettings`,
  when(ctx`hasSettings`, heroSetting)`settings`
)

export interface AdditionalInfo {
  winCondition: WinCondition
  winConditionData: WinConditionData
  loseCondition: LoseCondition
  loseConditionData: LoseConditionData
  teamsCount: number
  teams?: number[]
  availableHeroes: number[]
  customHeroesCount?: number
  customHeroes: CustomHero[]
  availableArtifacts: number[]
  availableSpells: number[]
  availableSkills: number[]
  rumorsCount: number
  rumors: Rumor[]
  heroSettings: HeroSettings[]
}

export const additionalInfo = struct(
  enums(uint8, winConditionsEnum)`winCondition`,
  branch(ctx`winCondition`, winConditionsBranches)`winConditionData`,
  enums(uint8, loseConditionsEnum)`loseCondition`,
  branch(ctx`loseCondition`, loseConditionsBranches)`loseConditionData`,
  uint8`teamsCount`,
  when(ctx`teamsCount`.gt(0), array(byte, H3M_MAX_PLAYERS))`teams`,
  bytes(context => (isRoE(context) ? 16 : 20))`availableHeroes`,
  when(isNotRoE, uint32),
  when(isSoD, uint8)`customHeroesCount`,
  when(isSoD, array(customHero, ctx`customHeroesCount`))`customHeroes`,
  skip(31),
  when(isNotRoE, bytes(context => (isSoD(context) ? 18 : 17)))`availableArtifacts`,
  when(isSoD, bytes(9))`availableSpells`,
  when(isSoD, bytes(4))`availableSkills`,
  uint32`rumorsRount`,
  array(rumor, ctx`rumorsRount`)`rumors`,
  when(isSoD, array(heroSettings, 156))`heroSettings`
)
