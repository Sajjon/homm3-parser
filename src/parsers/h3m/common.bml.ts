import {
  byte,
  uint8,
  int8,
  int16,
  uint32,
  int32,
  when,
  struct,
  uint16,
  array,
  pascalString,
  enums,
  ctx,
  Context,
} from 'binary-markup'

import { MapFormat } from './constants'
import { artifactEnum } from './enums/artifact'
import { creatureEnum } from './enums/creature'
import { heroEnum } from './enums/hero'
import { secondarySkillEnum, secondarySkillMasteryEnum } from './enums/skill'
import { spellEnum } from './enums/spell'
import { resourceEnum } from './enums/resource'
import { playerEnum } from './enums/player'
import {
  PrimarySkills,
  CreatureSlot,
  WornArtifacts,
  Backpack,
  Artifacts,
  SecondarySkillData,
} from './contracts/common'

export const isRoE = (context: Context): boolean => context.get('format') === MapFormat.ROE
export const isSoD = (context: Context): boolean => context.get('format') === MapFormat.SOD
export const isNotRoE = (context: Context): boolean => context.get('format') !== MapFormat.ROE

export const hommString = pascalString(uint32)

export const primarySkills = struct<PrimarySkills>(
  uint8`attack`,
  uint8`defense`,
  uint8`spellPower`,
  uint8`knowledge`,
)

export const skill = enums(byte, secondarySkillEnum)

export const secondarySkill = struct<SecondarySkillData>(
  skill`type`,
  enums(uint8, secondarySkillMasteryEnum)`level`,
)
export const experience = uint32
export const heroType = uint8
export const artifactType = when(isRoE, uint8, uint16)
export const creatureType = when(isRoE, int8, int16)
export const resourceType = uint8
export const spellType = uint8
export const absodId = uint32
export const playerFlag = enums(uint8, playerEnum)

export const artifact = enums(artifactType, artifactEnum)
export const creature = enums(creatureType, creatureEnum)
export const hero = enums(heroType, heroEnum)
export const spell = enums(spellType, spellEnum)
export const resource = enums(resourceType, resourceEnum)

export const creatureSlot = struct<CreatureSlot>(creature`type`, uint16`quantity`)

export const army = array(creatureSlot, 7)

export const resources = array(int32, 7)

export const wornArtifacts = struct<WornArtifacts>(
  artifact`headwear`,
  artifact`shoulders`,
  artifact`rightHand`,
  artifact`leftHand`,
  artifact`torso`,
  artifact`rightRing`,
  artifact`leftRing`,
  artifact`feet`,
  artifact`misc1`,
  artifact`misc2`,
  artifact`misc3`,
  artifact`misc4`,
  artifact`device1`,
  artifact`device2`,
  artifact`device3`,
  artifact`device4`,
  when(isSoD, uint16),
  artifact`spellbook`,
  artifact`misc5`,
)

export const backpack = struct<Backpack>(
  //
  uint16`count`,
  array(artifact, ctx`count`)`artifacts`,
)

export const artifacts = struct<Artifacts>(wornArtifacts`worn`, backpack`backpack`)

export const owner = enums(byte, playerEnum)
