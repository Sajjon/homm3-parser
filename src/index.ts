import { parse, StreamInput, TagOrWrapper } from 'binary-markup'
import * as pako from 'pako'
import { h3mFile } from './parsers/h3m/h3m.bml'
import { H3MFile } from './parsers/h3m/contracts'

export * from './parsers/assets'
export * from './parsers/h3m/contracts'
export * from './parsers/h3m/constants'

export function parseH3M(data: StreamInput): H3MFile {
  return parse<H3MFile>(h3mFile, data)
}

export function parseH3MFile(data: Uint8Array): H3MFile {
  const content = pako.inflate(data)
  return parseH3M(content.buffer)
}

export function parseAsset<T>(type: TagOrWrapper<T>, data: StreamInput): T {
  return parse(type, data)
}
