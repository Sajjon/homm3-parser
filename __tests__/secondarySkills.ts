import { testMap } from './utils'
import { H3MFile, MetaType } from '../src'
import { MapFormat } from '../src/parsers/h3m/constants'

describe('secondary scills', () => {
  testMap('secondary skills', 'secondary-skills.h3m', (map: H3MFile) => {
    if (map.format === MapFormat.SOD) {
      return map.objectDetails.entries
        .map(obj => {
          if (obj.body.type === MetaType.Hero) {
            const heroData = obj.body.data
            if (heroData.hasSecondarySkills) return heroData.secondarySkills
          }
          return null
        })
        .filter(Boolean)
    }
  })
})
