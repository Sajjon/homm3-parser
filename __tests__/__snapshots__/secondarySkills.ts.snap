// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`secondary scills should parse secondary skills 1`] = `
Array [
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Pathfinding",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Archery",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Logistics",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Scouting",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Diplomacy",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Navigation",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Leadership",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Wisdom",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Mysticism",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Luck",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Ballistics",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "EagleEye",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Necromancy",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Estates",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "FireMagic",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "AirMagic",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "WaterMagic",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "EarthMagic",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Scholar",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Tactics",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Artillery",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Learning",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Offense",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Armorer",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Intelligence",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Sorcery",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "Resistance",
      },
    ],
  },
  Object {
    "count": 1,
    "skills": Array [
      Object {
        "level": "Basic",
        "type": "FirstAid",
      },
    ],
  },
]
`;
